### PR TITLE
Enhancement: Improved L1 Verifier Error Handling

### PR DESCRIPTION
**Pull request by aminsato**
#1385

### CODE REVIEW

File: l1/l1.go [hunk 1]
```diff
@@ -78,7 +78,9 @@ func (c *Client) subscribeToUpdates(ctx context.Context, updateChan chan *contra
 func (c *Client) checkChainID(ctx context.Context) error {
 	gotChainID, err := c.l1.ChainID(ctx)
 	if err != nil {
-		return fmt.Errorf("retrieve Ethereum chain ID: %w", err)
+		// Instead of returning an error, log a warning and keep trying.
+		c.log.Warnw("Failed to retrieve Ethereum chain ID: ", err)
+		return nil
```
Comment (joshklop) [inline l1/l1.go:83]: We need to keep trying (i.e., no `return nil`). We'll need to put this in a loop and backoff when we receive an error.

**Review by joshklop — CHANGES_REQUESTED** [review]

**Comment by aminsato** [comment]
It still hasn't been reviewed by someone.
@omerfirmak 
@joshklop

File: l1/l1.go [hunk 2]
```diff
@@ -76,19 +76,28 @@ func (c *Client) subscribeToUpdates(ctx context.Context, updateChan chan *contra
 }
 
 func (c *Client) checkChainID(ctx context.Context) error {
-	gotChainID, err := c.l1.ChainID(ctx)
-	if err != nil {
-		return fmt.Errorf("retrieve Ethereum chain ID: %w", err)
-	}
+	retryAttempts := 3
+
+	for i := 0; i < retryAttempts; i++ {
```
Comment (joshklop) [inline l1/l1.go:81]: We should retry forever (never fail).

File: l1/l1.go [hunk 3]
```diff
@@ -76,19 +76,28 @@ func (c *Client) subscribeToUpdates(ctx context.Context, updateChan chan *contra
...
+		gotChainID, err := c.l1.ChainID(ctx)
+		if err != nil {
+			c.log.Warnw("Failed to retrieve Ethereum chain ID: ", err)
+			backOffDuration := time.Duration(i+1) * time.Second
+			time.Sleep(backOffDuration)
```
Comment (joshklop) [inline l1/l1.go:86]: The backoff should be exponential.

**Review by joshklop — CHANGES_REQUESTED** [review]

File: l1/l1.go [hunk 4]
```diff
@@ -76,19 +76,25 @@ func (c *Client) subscribeToUpdates(ctx context.Context, updateChan chan *contra
...
 func (c *Client) checkChainID(ctx context.Context) error {
-	gotChainID, err := c.l1.ChainID(ctx)
-	if err != nil {
-		return fmt.Errorf("retrieve Ethereum chain ID: %w", err)
-	}
+	for {
+		backOffDuration := time.Duration(1) * time.Second
+		gotChainID, err := c.l1.ChainID(ctx)
+		if err != nil {
+			c.log.Warnw("Failed to retrieve Ethereum chain ID: ", err)
```
Comment (joshklop) [inline l1/l1.go:83]:
```suggestion
			c.log.Warnw("Failed to retrieve Ethereum chain ID, retrying", err)
```
Let's move this below the backoff calculation and add `"waitFor", backoffDuration` to the log so the user knows how long Juno will wait.

File: l1/l1.go [hunk 5]
```diff
@@ -76,19 +76,25 @@ func (c *Client) subscribeToUpdates(ctx context.Context, updateChan chan *contra
...
+			backOffDuration *= 2
+			time.Sleep(backOffDuration)
```
Comment (joshklop) [inline l1/l1.go:85]: These lines should be swapped.

File: l1/l1.go [hunk 6]
```diff
@@ -76,19 +76,25 @@ func (c *Client) subscribeToUpdates(ctx context.Context, updateChan chan *contra
 }
 
 func (c *Client) checkChainID(ctx context.Context) error {
-	gotChainID, err := c.l1.ChainID(ctx)
-	if err != nil {
-		return fmt.Errorf("retrieve Ethereum chain ID: %w", err)
-	}
+	for {
+		backOffDuration := time.Duration(1) * time.Second
```
Comment (joshklop) [inline l1/l1.go:80]: This should go outside the `for` loop, otherwise we reset the backoff every iteration.

**Review by joshklop — CHANGES_REQUESTED** [review]

**Comment by joshklop** [comment]
We can wait for https://github.com/NethermindEth/juno/pull/1399 to be merged so we don't have to calculate backoffs and sleep manually.

File: l1/l1.go [hunk 7]
```diff
@@ -76,19 +76,25 @@ func (c *Client) subscribeToUpdates(ctx context.Context, updateChan chan *contra
 }
 
 func (c *Client) checkChainID(ctx context.Context) error {
-	gotChainID, err := c.l1.ChainID(ctx)
-	if err != nil {
-		return fmt.Errorf("retrieve Ethereum chain ID: %w", err)
-	}
+	backOffDuration := time.Duration(1) * time.Second
```
Comment (kirugan) [inline l1/l1.go:79]: nit: `backOffDuration := time.Second`

**Pull request closed without merging**